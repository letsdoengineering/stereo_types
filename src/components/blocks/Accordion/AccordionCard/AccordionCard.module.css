.head-section--bottom-border {
    border-bottom: 1px solid var(--outline-gray);
    margin-bottom: -1px;
}

.accordion {
    cursor: pointer;
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    width: 100%;
    font: inherit;
    outline: inherit;
}

button.accordion:focus {
    outline: 5px auto Highlight;
    outline: 5px auto -webkit-focus-ring-color;
}

.accordion-content {
    background-color: inherit;
    overflow: hidden;
    transition: max-height 0.1s ease;
}

.icon {
    margin: 0 0.5rem 0 0.5rem;
    right: 0;
}

.icon-container {
    position: relative;
}

.rotate {
    transform: rotate(180deg);
}
