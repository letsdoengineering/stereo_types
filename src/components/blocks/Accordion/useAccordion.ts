import React, { useRef, useState } from 'react'
import * as self from './useAccordion'

type UseAccordion = {
  /** indicates whether the accordion should be open or not */
  isOpen: boolean
  /** used to set isOpen state */
  setIsOpen: React.Dispatch<React.SetStateAction<boolean>>
  /** value of the scrollHeight of the div containing the accordion's content
   *  e.g. '10px' used to expand the accordion */
  scrollHeight: string
  /** used to set the scrollHeight state */
  setScrollHeight: React.Dispatch<React.SetStateAction<string>>
  /** a React ref for use with the expandable/collapsable element in the accordion */
  ref: React.RefObject<HTMLDivElement>
  /** sets isOpen state and scrollHeight */
  toggleAccordion(): void
}

/** getScrollHeight: returns a string indicating scrollHeight of an element based on whether it is active or not
 * @param {boolean} isActive - indicates whether to return 0px or ${scrollHeight}px
 * @param {number} scrollHeight - the scrollHeight of an element - e.g. as taken from an element's ref */
export const getScrollHeight = (isActive: boolean, scrollHeight: number | undefined): string => {
  return isActive ? '0px' : `${scrollHeight}px`
}

/** useAccordion: custom hook returns an object containing state for use with
 * accordion components and a toggleAccordion function
 * @returns { object } UseAccordion */
export default (): UseAccordion => {
  const [isOpen, setIsOpen] = useState(false)
  const [scrollHeight, setScrollHeight] = useState('0px')
  const ref = useRef<HTMLDivElement>(null)

  const toggleAccordion = (): void => {
    setIsOpen(!isOpen)
    setScrollHeight(self.getScrollHeight(isOpen, ref?.current?.scrollHeight))
  }

  return {
    isOpen,
    setIsOpen,
    scrollHeight,
    setScrollHeight,
    ref,
    toggleAccordion,
  }
}
