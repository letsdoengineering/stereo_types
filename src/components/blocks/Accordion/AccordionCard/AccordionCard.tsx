import React from 'react'
import Card from '../../Card/Card'
import GridRow from '../../../basics/Grid/GridRow/GridRow'
import * as accordionCardStyles from './AccordionCard.module.css'
import getClassNames from '../../../../utils/get-class-names'
import GridColumn from '../../../basics/Grid/GridColumn/GridColumn'
import useAccordion from '../useAccordion'
import IconButton from '../../../basics/IconButton/IconButton'

type AccordionCardProps = {
  /** JSX passed in to render in the top of the card displayed before the accordion content */
  headSection: React.ReactNode
  /** JSX passed in for the accordion content itslef */
  innerContent: React.ReactNode
  /** a baseId used to create a number of unique IDs and associated attributes */
  baseId: string
}

/** AccordionCard: returns a Card that behaves like an accordion */
const AccordionCard: React.FC<AccordionCardProps> = ({
  headSection,
  innerContent,
  baseId,
}: AccordionCardProps) => {
  const { isOpen, scrollHeight, toggleAccordion, ref } = useAccordion()

  const contentId = `${baseId}-content`
  const accordionId = `${baseId}-accordion`

  const iconClasses = getClassNames({
    defaultClasses: [accordionCardStyles.icon],
    conditionalClasses: {
      [accordionCardStyles.rotate]: isOpen,
    },
  })

  return (
    <Card>
      <GridRow className={accordionCardStyles.headSectionBottomBorder}>
        <GridColumn size='11'>{headSection}</GridColumn>
        <GridColumn>
          <IconButton
            icon='chevron-down'
            iconSize='XL'
            iconClassName={iconClasses}
            buttonClassName={accordionCardStyles.accordion}
            label={accordionId}
            aria-expanded={isOpen}
            id={accordionId}
            aria-controls={contentId}
            iconColor='Blue'
            onClick={toggleAccordion}
          />
        </GridColumn>
      </GridRow>
      <div
        id={contentId}
        aria-labelledby={accordionId}
        ref={ref}
        role='region'
        style={{ maxHeight: `${scrollHeight}` }}
        className={accordionCardStyles.accordionContent}
      >
        {innerContent}
      </div>
    </Card>
  )
}

export default AccordionCard
