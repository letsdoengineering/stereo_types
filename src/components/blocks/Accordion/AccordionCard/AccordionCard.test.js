import * as React from 'react'
import { render, screen } from '@testing-library/react'
import AccordionCard from './AccordionCard'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'
import * as useAccordion from '../useAccordion'
import userEvent from '@testing-library/user-event'

describe('AccordionCard', () => {
  it('has not changed the JSX layout of the page from previous snapshot', () => {
    const tree = renderer
      .create(
        <AccordionCard headSection={<div>something</div>} innerContent={<div>something</div>} />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should adjust the scrollheight of the content container when user clicks the accordion button', () => {
    const getScrollHeightSpy = jest.spyOn(useAccordion, 'getScrollHeight')
    render(
      <AccordionCard
        baseId='test'
        headSection={<div>some header</div>}
        innerContent={<div>some content</div>}
      />
    )

    userEvent.click(screen.getByText('test-accordion'))
    expect(getScrollHeightSpy).toHaveBeenCalledTimes(1)
  })

  it('renders the page DOM that is accessible', async () => {
    const { container } = render(
      <AccordionCard headSection={<a>something</a>} innerContent={<div>something</div>} />
    )
    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
