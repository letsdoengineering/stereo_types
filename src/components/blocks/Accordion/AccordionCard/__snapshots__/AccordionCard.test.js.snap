// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionCard has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  className="card"
>
  <div
    className="gridRow headSectionBottomBorder"
  >
    <div
      className="gridColumn size11"
    >
      <div>
        something
      </div>
    </div>
    <div
      className="gridColumn sizeAuto"
    >
      <button
        aria-controls="undefined-content"
        aria-expanded={false}
        className="button accordion"
        disabled={false}
        id="undefined-accordion"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden="true"
          className="fas fa-chevron-down iconSizeXL iconColorBlue buttonIcon icon"
        />
        <span
          className="visually-hidden"
        >
          undefined-accordion
        </span>
      </button>
    </div>
  </div>
  <div
    aria-labelledby="undefined-accordion"
    className="accordionContent"
    id="undefined-content"
    role="region"
    style={
      Object {
        "maxHeight": "0px",
      }
    }
  >
    <div>
      something
    </div>
  </div>
</div>
`;
